solarforecastarbiter/metrics/deterministic.py

```diff
@@ -545,6 +549,151 @@ def combined_performance_index(obs, fx):
+    agg_fnc = _np_agg_fnc(cost_params.aggregation, cost_params.net)
+
+    errors = error_fnc(obs, fx)
+    return agg_fnc(errors) * cost_const
+
+
+def _make_time_of_day_cost_ser(times, costs, index, tz, fill):
+    if len(index) == 0 or len(times) == 0:
+        return 0
+    dates = list(np.unique(index.date))
+    # extend dates +- 1 day so that index is within the cost
+    # ser we construct
+    dates.insert(0, min(dates) - dt.timedelta(days=1))
+    dates.insert(-1, max(dates) + dt.timedelta(days=1))
+
+    # insert the last cost at 00:00 if not present so forward
+    # fill works (sometimes one date w/ tz adjust not enough)
+    if fill == 'ffill' and dt.time(0) not in times:
+        max_ind = np.argmax(times)
+        times.insert(0, dt.time(0))
+        costs.insert(0, costs[max_ind])
+    # insert the first cost at 23:59 if not present so back
+    # fill works even if tz adjusts index
+    elif fill == 'bfill' and dt.time(23, 59, 59) not in times:
+        min_ind = np.argmin(times)
+        times.insert(-1, dt.time(23, 59, 59))
+        costs.insert(-1, costs[min_ind])
+    # make the cost series
+    prod = [(pd.Timestamp.combine(x, y[0]), y[1])
+            for x in dates for y in zip(times, costs)]
+    base_ser = pd.DataFrame(
+        prod, columns=['timestamp', 'cost']
+    ).set_index('timestamp')['cost'].tz_localize(tz).sort_index()
+    # only get those values at index filling as appropriate
+    ser = base_ser.reindex(index, method=fill)
+    return ser
+
+
+def _time_of_day_cost(obs, fx, cost_params, error_fnc):
+    """Compute cost according to a datamodel.TimeOfDayCost"""
+    agg_fnc = _np_agg_fnc(cost_params.aggregation, cost_params.net)
+    fill = _FILL_OPTIONS[cost_params.fill]
+
+    errors = error_fnc(obs, fx)
+    tz = cost_params.timezone or errors.index.tzinfo
+    cost_ser = _make_time_of_day_cost_ser(
+        cost_params.times, cost_params.costs, errors.index, tz, fill)
+    error_cost = errors * cost_ser
+    return agg_fnc(error_cost)
+
+
+def _datetime_cost(obs, fx, cost_params, error_fnc):
+    """Compute cost according to a datamodel.DatetimeCost"""
+    agg_fnc = _np_agg_fnc(cost_params.aggregation, cost_params.net)
+    fill = _FILL_OPTIONS[cost_params.fill]
+    cost_ser = pd.Series(cost_params.costs,
+                         index=pd.DatetimeIndex(cost_params.datetimes),
+                         dtype=float)
+
+    errors = error_fnc(obs, fx)
+
+    if cost_ser.index.tzinfo is None:
+        if cost_params.timezone is not None:
+            cost_ser = cost_ser.tz_localize(cost_params.timezone)
+        else:
+            cost_ser = cost_ser.tz_localize(errors.index.tzinfo)
+
+    error_cost = errors * cost_ser.reindex(errors.index, method=fill)
+    return agg_fnc(error_cost)
+
+
+def _band_masks(bands, errors):
+    """Make masks for each band based on which band errors falls in"""
+    prev = np.zeros(errors.shape, dtype=bool)
+    out = []
+    for band in bands:
```

Comment:
can this be replaced with [`np.digitize`](https://numpy.org/doc/stable/reference/generated/numpy.digitize.html)?

Reply:
see the comment about the range endpoint in/exclusion